### PR TITLE
Fix latch

### PR DESCRIPTION
**Pull request by msimberg**
Add `notified_` member variable to `latch` to avoid `wait` returning before `count_down` has notified other threads. Details on the race in the commit message.

This seems to be the cause of the segfaults we've seen with the parallel algorithms after #3745.

This was the best I could do for now for avoiding taking the lock. It's still only taken twice: once for the notify and once when waiting. Ideas for avoiding that while keeping the behavior correct are welcome.

### CODE REVIEW

File: hpx/lcos/local/latch.hpp [hunk 1]
```diff
@@ -134,12 +138,9 @@ namespace hpx { namespace lcos { namespace local
         ///
         void wait() const
         {
-            if (counter_.load(std::memory_order_acquire) > 0)
-            {
-                std::unique_lock<mutex_type> l(mtx_.data_);
-                if (counter_.load(std::memory_order_relaxed) > 0)
-                    cond_.data_.wait(l, "hpx::local::latch::wait");
-            }
+            std::unique_lock<mutex_type> l(mtx_.data_);
```
Comment (hkaiser) [inline hpx/lcos/local/latch.hpp]: Can we do a double-test still? I.e. once before grabbing the lock, just testing the counter and then, if necessary grab the lock and perform a repeated test under the lock?

Comment (msimberg) [reply hpx/lcos/local/latch.hpp]: Hmm, don't think so. Just testing `counter_ > 0` outside the lock would get us to the same situation as before. Testing `counter_ > 0 || !notified` is significantly less racey, but still racey. Do you mean something else?

Comment (hkaiser) [reply hpx/lcos/local/latch.hpp]: Makes sense, thanks!

**Comment by msimberg** [comment]
Hmm... there's more to this: http://cdash.cscs.ch/testDetails.php?test=7767107&build=50026. I also need to fix the assert in `reset`, so no merging, please.

**Comment by msimberg** [comment]
@hkaiser after these fixes I can't reproduce the segfaults anymore, even though I've seen two happen on pycicle. I think this should still be a strict improvement (at least in correctness, hopefully not a big regression in performance) and try to figure out if more segfaults still happen on rostam. Anything against merging this?

**Comment by hkaiser** [comment]
> Anything against merging this?

@msimberg please go ahead.

**Pull request merged**